client/game_logic.py

```diff
@@ -0,0 +1,190 @@
+                self.values_changed = False
+                self.whack_moles()
+                self.generate_new_moles()
+                self.moles_progression()
+                if self.values_changed and self.last_print_time + 1 < time.time():
+                    self.print_board()
+                    self.package_number += 1
+                    self.last_print_time = time.time()
+                    if self.score >= MAX_SCORE:
+                        self.end = True
+                    print(self.create_bits_package())
+                time.sleep(0.2)
+        except KeyboardInterrupt:
+            return
+
+    def generate_mole_lifetime(self) -> list:
+        value = random.uniform(self.mole_life_time[0], self.mole_life_time[1])
+        return [time.time() + value * (x + 1) / 3 for x in range(3)]
+
+    def set_new_mole(self, pos: int):
+        self.moles_status[pos] = [3, self.generate_mole_lifetime()]
+
+    def hide_mole(self, pos: int):
+        self.moles_status[pos] = [0, 0]
+
+    def moles_progression(self):
+        self.moles_lock.acquire()
+        op_start_time = time.time()
+        for idx, mole in enumerate(self.moles_status):
+            for state, checked_time in enumerate(mole[1][::-1]):
+                if checked_time < op_start_time:
+                    self.moles_status[idx][0] = state
+                    self.values_changed = True
+                    break
+        self.moles_lock.release()
+
+    def generate_new_moles(self):
+        self.moles_lock.acquire()
+        while sum([bool(x[0]) for x in self.moles_status]) < self.max_moles_amount:
+            rand = random.random()
+            if self.moles_status[int(rand * self.number_of_moles)][0] == 0:
+                self.set_new_mole(int(rand * self.number_of_moles))
+                self.values_changed = True
+        self.moles_lock.release()
+
+    def whack_moles(self):
+        self.pressed_keys_lock.acquire()
+        unique_keys = []
+        for key in self.pressed_key:
+            if key not in unique_keys:
+                unique_keys.append(key)
+        self.pressed_key = []
+        self.pressed_keys_lock.release()
+        self.moles_lock.acquire()
+        for key in unique_keys:
+            self.values_changed = True
+            self.calculate_score(key)
+        self.moles_lock.release()
+
+    def calculate_score(self, key: int, max_combo=8):
+        if sum(self.moles_status[key]) != 0:
+            self.score += self.combo
+            self.combo *= 2 if self.combo < max_combo else 1
+            self.hide_mole(key)
+        else:
+            self.combo = 1
+
+    def print_board(self):
+        self.moles_lock.acquire()
+        row_length = int(math.sqrt(self.number_of_moles))
+        for idx in range(row_length):
+            print("\t".join([str(x[0])
+                             for x in self.moles_status[idx * row_length:(idx + 1) * row_length]]))
+        print("\n")
+        self.moles_lock.release()
```

Comment:
It's just for debugging purpose for now, right?

Reply:
Yes, i'd suggest to remove it as soon as you will be totally sure everything here works fine